File: app/src/main/java/com/zofers/zofers/model/Conversation.java
package com.zofers.zofers.model;

import android.content.Context;

import com.google.common.base.Strings;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

public class Conversation {
	public static final String DOC_NAME = "conversation";
	private String id;
	private String name;
	private Date creationDate;
	private Date lastActionDate;

	private List<String> participantIDs;
	private List<Participant> participants;

	private Message lastMessage;


	public String getId() {
		return id;
	}

	public void setId(String id) {
		this.id = id;
	}

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public Date getCreationDate() {
		return creationDate;
	}

	public void setCreationDate(Date creationDate) {
		this.creationDate = creationDate;
	}

	public Date getLastActionDate() {
		return lastActionDate;
	}

	public void setLastActionDate(Date lastActionDate) {
		this.lastActionDate = lastActionDate;
	}

	public List<Participant> getParticipants() {
		return participants;
	}

	public void setParticipants(List<Participant> participants) {
		this.participants = participants;
	}

	public Message getLastMessage() {
		return lastMessage;
	}

	public void setLastMessage(Message lastMessage) {
		this.lastMessage = lastMessage;
	}

	public List<String> getParticipantIDs() {
		return participantIDs;
	}

	public void setParticipantIDs(List<String> participantIDs) {
		this.participantIDs = participantIDs;
	}

	//_____________________________________________________________________

	public Participant getUser(String id) {
		for (Participant u : participants) {
			if (u.getId().equals(id)) {
				return u;
			}
		}
		return null;
	}

	public List<Participant> getParticipantsExcept(String userId) {
		List<Participant> returnList = new ArrayList<>();
		for (Participant p : participants) {
			if (!p.getId().equals(userId)) {
				returnList.add(p);
			}
		}
		return returnList;
	}

	public Participant getParticipant(String id) {
		for (Participant p : participants) {
			if (p.getId().equals(id)) {
				return p;
			}
		}
		return null;
	}

	public void addParticipant(Participant participant) {
		if (participants == null) {
			participants = new ArrayList<>();
		}
		participants.add(participant);
	}

	public void removeParticipant(Participant user) {
		for (Participant p : participants) {
			if (p.getId().equals(user.getId())) {
				break;
			}
		}

	}

	public boolean isGroup() {
		return participants.size() > 2;
	}

	public String getDisplayName(String userId, Context context) {

		if (Strings.isNullOrEmpty(name)) {
			return name;
		}
		if (!isGroup()) {
			return getParticipantsExcept(userId).get(0).getName();
		}
		String displayName = "";
		int activeUsersCount = 0;
		Participant me = null;
		for (int i = 0; i < participants.size(); i++) {
			Participant user = participants.get(i);
			if (user.getId().equals(userId)) {
				me = user;
				continue;
			}
			if (activeUsersCount < 3) { // first 3 names
				if (Strings.isNullOrEmpty(displayName)) {
					displayName = user.getName();
				} else {
					displayName += ", " + user.getName();
				}
			}
			activeUsersCount++;
		}
		if (activeUsersCount == 0 && me != null) { // group chat, where only me
			displayName = me.getName();
		} else if (activeUsersCount > 3) {
			displayName += " ..."; //String.format(context.getString(R.string.messenger_group_name_others), (activeUsersCount - 3));
		} else if (activeUsersCount == 0) { // me == null
			displayName = "";
		}
		return displayName;
	}

	public boolean isUnread(String userId) {
		Participant participant = getParticipant(userId);
		if (participant == null) return false;
		if (lastMessage != null) {
			return !userId.equals(lastMessage.getUserId()) && !lastMessage.getId().equals(participant.getLastSeenMessageId());
		} else {
			return participant.getLastSeenMessageId() == null;
		}
	}

}

File: app/src/main/java/com/zofers/zofers/model/Conversation.kt
package com.zofers.zofers.model

import java.util.*

class Conversation {

	var id: String? = null
	var name: String? = null
	var creationDate: Date? = null
	var lastActionDate: Date? = null
	var participantIDs: List<String>? = null
	var participants: List<Participant>? = null
	var lastMessage: Message? = null
	//todo can be removed if not used by IOS
	var status: Int? = 0 // requested - 0, approved - 1, rejected - 2

	//_____________________________________________________________________
	fun getUser(id: String): Participant? {
		for (u in participants!!) {
			if (u.id == id) {
				return u
			}
		}
		return null
	}

	fun getParticipantsExcept(userId: String): List<Participant> {
		val returnList: MutableList<Participant> = ArrayList()
		for (p in participants!!) {
			if (p.id != userId) {
				returnList.add(p)
			}
		}
		return returnList
	}

	fun getParticipant(id: String): Participant? {
		for (p in participants!!) {
			if (p.id == id) {
				return p
			}
		}
		return null
	}

	fun removeParticipant(user: Participant) {
		for (p in participants!!) {
			if (p.id == user.id) {
				break
			}
		}
	}

	val isGroup: Boolean
		get() = participants!!.size > 2

	fun isUnread(userId: String): Boolean {
		val participant = getParticipant(userId) ?: return false
		return if (lastMessage != null) {
			userId != lastMessage!!.userID && lastMessage!!.id != participant.lastSeenMessageID
		} else {
			participant.lastSeenMessageID == null
		}
	}

	companion object {
		const val DOC_NAME = "conversation"

		const val STATUS_REQUESTED = 0
		const val STATUS_ACCEPTED = 1
		const val STATUS_REJECTED = 2
	}
}